### PR TITLE
3.x Implementation of spark bigquery connector with common library

### PR DESCRIPTION
**Pull request by BhagyaaShreeS**
Modules to be considered: spark-bigquery-dsv2/spark-bigquery-dsv2-common; spark-bigquery-dsv2/spark-bigquery-spark3

### CODE REVIEW

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryIndirectDataSourceWriter.java [hunk 1]
```diff
@@ -61,19 +45,11 @@
  * format, and then triggering a BigQuery load job on this data. Hence the "indirect" - the data
  * goes through an intermediate storage.
  */
-public class BigQueryIndirectDataSourceWriter implements DataSourceWriter {
+public class BigQueryIndirectDataSourceWriter implements DataSourceWriter, Serializable {
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-spark24/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryIndirectDataSourceWriter.java:48]: Why it needs to be Serializable?

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/v2/JavaShakespeare.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
+package com.google.cloud.spark.bigquery.v2;
+
+import java.io.PrintStream;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+
+public class JavaShakespeare {
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/v2/JavaShakespeare.java:8]: Please delete this class

File: spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/v2/WriteTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+package com.google.cloud.spark.bigquery.v2;
+
+import com.google.cloud.spark.bigquery.integration.model.Friend;
+import com.google.cloud.spark.bigquery.integration.model.Link;
+import com.google.cloud.spark.bigquery.integration.model.Person;
+import java.util.Arrays;
+import org.apache.spark.sql.*;
+
+public class WriteTest {
```
Comment (davidrabinowitz) [inline spark-bigquery-dsv2/spark-bigquery-spark24/src/test/java/com/google/cloud/spark/bigquery/v2/WriteTest.java:9]: Please use integration tests instead

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
* Needs to be rebased on current master
* CLAs need to be signed by all committers
* Please remove the spark3 implementation
* The spark 2.4 implementation should be shell wrapping the common code, it should not contain extra logic

Does not compile:
```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /Users/davidrab/dev/workspaces/dataproc/spark-bigquery-connector-maven/spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBigQueryDataSourceReader.java:[279,5] cannot find symbol
  symbol:   class List
  location: class com.google.cloud.spark.bigquery.common.GenericBigQueryDataSourceReader
[ERROR] /Users/davidrab/dev/workspaces/dataproc/spark-bigquery-connector-maven/spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBigQueryDataSourceReader.java:[279,39] cannot find symbol
  symbol:   class ArrayList
  location: class com.google.cloud.spark.bigquery.common.GenericBigQueryDataSourceReader
[ERROR] /Users/davidrab/dev/workspaces/dataproc/spark-bigquery-connector-maven/spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBigQueryDataSourceReader.java:[280,5] cannot find symbol
  symbol:   class List
  location: class com.google.cloud.spark.bigquery.common.GenericBigQueryDataSourceReader
[ERROR] /Users/davidrab/dev/workspaces/dataproc/spark-bigquery-connector-maven/spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBigQueryDataSourceReader.java:[280,41] cannot find symbol
  symbol:   class ArrayList
  location: class com.google.cloud.spark.bigquery.common.GenericBigQueryDataSourceReader
```

**Review by davidrabinowitz — CHANGES_REQUESTED** [review]
Please make sure that all files have the license header

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    <parent>
+        <groupId>com.google.cloud.spark</groupId>
+        <artifactId>spark-bigquery-parent</artifactId>
+        <version>0.30.0-SNAPSHOT</version>
+        <relativePath>../../spark-bigquery-parent</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>spark-bigquery-dsv2-common</artifactId>
+    <name>DSV2 BigQuery Connector Common Library</name>
+    <licenses>
```
[the remark this answers is not in the document]
Comment (BhagyaaShreeS) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml:14]: Sure, David. I will have it removed.

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>com.google.cloud.spark</groupId>
+        <artifactId>spark-bigquery-parent</artifactId>
+        <version>0.30.0-SNAPSHOT</version>
+        <relativePath>../../spark-bigquery-parent</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+    <artifactId>spark-bigquery-dsv2-common</artifactId>
+    <name>DSV2 BigQuery Connector Common Library</name>
```
[the remark this answers is not in the document]
Comment (BhagyaaShreeS) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml:13]: I will modify the name

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
+            <distribution>repo</distribution>
+        </license>
+    </licenses>
+    <dependencies>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>bigquery-connector-common</artifactId>
+            <version>${project.version}</version>
+            <scope>compile</scope>
```
[the remark this answers is not in the document]
Comment (BhagyaaShreeS) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/pom.xml:26]: I will have it removed, David.

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBQDataSourceReaderHelper.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.google.cloud.spark.bigquery.common;
+
+import java.io.Serializable;
+
+public class GenericBQDataSourceReaderHelper implements Serializable {
```
[the remark this answers is not in the document]
Comment (BhagyaaShreeS) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBQDataSourceReaderHelper.java:5]: I will refactor the name everywhere

File: spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBQDataSourceReaderHelper.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+
+  public boolean isBatchReadEnable() {
```
[the remark this answers is not in the document]
Comment (BhagyaaShreeS) [reply spark-bigquery-dsv2/spark-bigquery-dsv2-common/src/main/java/com/google/cloud/spark/bigquery/common/GenericBQDataSourceReaderHelper.java:7]: I will have it refactored David

**Pull request closed without merging**